resources/fetchersManager/data.go

```diff
@@ -50,68 +57,87 @@ func NewData(log *logp.Logger, interval time.Duration, timeout time.Duration, fe
 		timeout:    timeout,
 		interval:   interval,
 		fetchers:   fetchers,
-		stop:       make(chan struct{}),
-		stopNotice: make(chan time.Duration),
+		stopNotice: make(chan time.Duration, 1),
 	}, nil
 }
 
+var ErrGracePeriodTooLong = errors.New("graceful shutdown too long")
+var ErrGracePeriodEnded = errors.New("graceful shutdown ended")
+
 // Run starts all configured fetchers to collect resources.
 func (d *Data) Run(ctx context.Context) Stop {
-	go d.fetchAndSleep(ctx)
-	called := atomic.NewBool(false)
-	return func(ctx context.Context, grace time.Duration) {
-		defer called.Store(true)
-		if called.Load() {
+	fetchCtx, cancel := context.WithCancelCause(ctx)
+	go d.fetchAndSleep(fetchCtx, cancel)
+
+	return func(grace time.Duration) {
+
+		d.stopNotice <- grace // Notify grace period started
+
+		now := time.Now()
+
+		select {
+
+		// Exit after grace period
+		case <-time.After(grace):
+			fmt.Printf("shutdown grace period ended after %s \n", time.Since(now).String())
+			cancel(ErrGracePeriodEnded)
+			return
+
+		// Exit before grace period
+		case <-fetchCtx.Done():
+			fmt.Printf("context canceled after %s \n", time.Since(now).String())
 			return
 		}
-		d.stopData(ctx, grace)
 	}
 }
 
-func (d *Data) fetchAndSleep(ctx context.Context) {
-	ctx, cancel := context.WithCancel(ctx)
-	// set immediate exec for first time run
-	timer := time.NewTimer(0)
+func (d *Data) fetchAndSleep(ctx context.Context, cancel context.CancelCauseFunc) {
+
+	interval := time.NewTimer(0) // Initial interval is 0 to start immediately
+
 	defer func() {
-		cancel()
-		timer.Stop()
+		interval.Stop()
+		d.fetchers.Stop()
 	}()
-	run := atomic.NewBool(true)
+
+	exitNextIteration := atomic.NewBool(false)
+
 	for {
 		select {
-		case grace := <-d.stopNotice:
-			d.log.Infof("Received stop notice with grace period of %s, fetchers will not be executing from now on", grace.String())
-			run.Store(false)
-		case <-d.stop:
-			d.log.Info("Fetchers manager stopped")
-			return
 		case <-ctx.Done():
-			d.log.Info("Fetchers manager canceled")
 			return
-		case <-timer.C:
-			if !run.Load() {
+
+		case grace := <-d.stopNotice:
+			fmt.Printf("shutdown grace period started. will exit after %s or next iteration \n", grace.String())
+			exitNextIteration.Store(true)
+
+		case <-interval.C:
+			if exitNextIteration.Load() {
+				fmt.Printf("exit new fetch iteration\n")
+				cancel(ErrGracePeriodTooLong)
 				return
```

Comment:
this is essentially our mechanism for shutting down before the grace period ended. 
it depends on the interval period, which in real envs is in mins/hrs, so really we'll only exit after waiting the full grace period.